components/places/src/storage/history.rs

```diff
@@ -210,6 +210,49 @@ pub fn delete_place_visit_at_time(db: &impl ConnExt, place: &Url, visit: Timesta
     Ok(())
 }
 
+pub fn prune_destructively(db: &impl ConnExt) -> Result<()> {
+    // For now, just fall back to wipe_local until we decide how this should work.
+    wipe_local(db)
+}
+
+pub fn wipe_local(db: &impl ConnExt) -> Result<()> {
+    let tx = db.unchecked_transaction()?;
+    wipe_local_in_tx(db)?;
+    tx.commit()?;
+    Ok(())
+}
+
+fn wipe_local_in_tx(db: &impl ConnExt) -> Result<()> {
+    use crate::frecency::DEFAULT_FRECENCY_SETTINGS;
+    db.execute_all(&[
+        "DELETE FROM moz_places
+         WHERE foreign_count == 0",
+        "DELETE FROM moz_historyvisits",
+        "DELETE FROM moz_places_tombstones",
+        "DELETE FROM moz_inputhistory",
+        "DELETE FROM moz_historyvisit_tombstones",
+        "DELETE FROM moz_origins
+         WHERE id NOT IN (SELECT origin_id FROM moz_places)",
+        &format!(
+            "UPDATE moz_places SET
+                frecency = {unvisited_bookmark_frec},
+                sync_change_counter = 0",
+            unvisited_bookmark_frec = DEFAULT_FRECENCY_SETTINGS.unvisited_bookmark_bonus
+        ),
+    ])?;
+
+    let need_frecency_update =
+        db.query_rows_and_then_named("SELECT id FROM moz_places", &[], |r| {
+            r.get_checked::<_, RowId>(0)
+        })?;
+    // Update the frecency for any remaining items, which basically means just
+    // for the bookmarks.
+    for row_id in need_frecency_update {
+        update_frecency(db.conn(), row_id, None)?;
+    }
+    Ok(())
```

Comment:
should we vacuum here?

Reply:
Not there (can't inside a transaction), but it's easy enough to do this outside.